05_testing_and_ci/python_testing_exercise.md

```diff
@@ -0,0 +1,102 @@
+- The functions `initialize_domain` and `initialize_physical_parameters` have default values for all input parameters, hence they can be called without any parameters.
+- The file also has a `main()` function which shows a step-by-step procedure to solve the diffusion problem using an object of the class `SolveDiffusion2D`.
+- Make sure that `NumPy` and `Matplotlib` is installed on the system that you are working on.
+- Try to run this example in the following way:
+
+```python
+python3 diffusion2d.py
+```
+
+- Observe the output produced, it should be the same output as what was seen during the packaging exercise.
+- Spend some time understanding what each function does. It is important to understand all the functions as you will be writing tests for these functions.
+
+## Step 2 - Adding assertion statements
+
+- Add assertion statements to the functions `initialize_domain` and `initialize_physical_parameters` which check whether all input parameter variables have type `float`.
+
+## Step 3 - Writing unit tests
+
+- In the repository there is a folder `tests/`. In this folder there are two folders `unit/` and `integration/`. The first tests written will be unit tests.
+- In the file `tests/unit/test_diffusion2d_functions.py` there is already a skeleton code for three unit tests which are to be implemented.
+- The name of each test is of the format `test_<name_of_function_being_tested>`.
+- As these are unit tests, in each test only the call to the respective function needs to be made. No other function from `diffusion2d.py` must be called.
+- In this step you will write all the tests in a way that they can be run using `pytest` and not using the `unittest` format.
+- As an example, let us look at how we can write a unit test for the function `initialize_domain`
+    - When the function `initialize_domain` is being tested, you need to first identify which variables are being calculated in this function.
+    - In this case they are the variables `nx` and `ny`. Now choose some values for the variables `w`, `h`, `dx`, and `dy` which are different from the default values. For these values manually calculate the `nx` and `ny`.
+    - These manually calculated values are the expected values in this test.
+    - Now call the function `initialize_domain` which the chosen values of `w`, `h`, `dx`, and `dy` and using check if the `nx` and `ny` in the class member variables is equal to your expected values.
+    - Note that you have the object of the class `SolveDiffusion2D` and hence you can access member variables, for example `solver.nx` and `solver.ny`. This is useful to check the actual values.
+    - This check is done using an assertion statement, something that you have already seen in the demo during the lecture.
+- Using a similar workflow, complete the other two unit tests.
```

Comment:
Both these tests are actually not integration tests. It is expected that the student will set values of member functions of the class `SolveDiffusion2D` explicitly and thereby circumvent the need to call any functions other than the function being tested. Was this somehow not clear here? I will try to state it more explicitly.

Reply:
To me it was not clear. Maybe it would be after the full lecture. I am not sure whether it is forbidden/bad practice to call more than one member function in the unit test to set up the object.

Reply:
I have observed that for unit tests you really call only that function and define the required variables manually. This often involves defining member variables of a class object which may look ugly but needs to be done.

Reply:
Then maybe add a remark about that?

Related question, but not needed for the exercise: What do you in languages like C++ where cannot simply manipulate (private) member variables of classes? Should be the class be a self-contained unit in this case?

Reply:
> Related question, but not needed for the exercise: What do you in languages like C++ where cannot simply manipulate (private) member variables of classes? Should be the class be a self-contained unit in this case?

Good question! I dont have an answer for this now, but lets discuss this before the C++ / Boost testing lecture. I am sure there is a solution for this.

Reply:
@uekerman possibly relevant point for the C++ testing part.